### PR TITLE
RzIL: replace float comparsion macros with functions

### PR DESCRIPTION
**Pull request by p371jah**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

FEQ, FNEQ, FLT, FLE, FGT, FGE macros are too bulky. Their use leads to memory leaks.
I guess new functions don't really fit to `il_opcodes.c` but idk where to place them.

### CODE REVIEW

**Review by XVilka — CHANGES_REQUESTED** [review]
There is `librz/il/il_routines.c` precisely for this purpose. Maybe it could also be renamed into `il_helpers.c` to be clearer, the file Doxygen description should be updated too.

**Comment by XVilka** [comment]
Note that it produces different results now:
```
[XX] db/asm/arm_32 <asm> vceq.f32 d0, d1, d2
-- <asm> vceq.f32 d0, d1, d2 <--- 020e01f2 ---> <IL>
-- IL
--- expected
+++ actual
@@ -1,1 +1,1 @@
-(seq empty (set d0 (<< (cast 64 false (ite (&& (! (|| (is_nan (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) )) (is_nan (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) )))) (! (|| (|| (is_nan (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) )) (is_nan (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) ))) (|| (<. (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) ) (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) )) (<. (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) ) (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) )))))) (~ (bv 32 0x0)) (bv 32 0x0))) (bv 8 0x0) false)) (set d0 (<< (cast 64 false (ite (&& (! (|| (is_nan (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) )) (is_nan (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) )))) (! (|| (|| (is_nan (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) )) (is_nan (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) ))) (|| (<. (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) ) (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) )) (<. (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) ) (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) )))))) (~ (bv 32 0x0)) (bv 32 0x0))) (bv 8 0x20) false)))
+(seq empty (set d0 (<< (cast 64 false (ite (&& (! (|| (is_nan (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) )) (is_nan (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) )))) (! (|| (<. (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) ) (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) )) (<. (float 0 (cast 32 false (>> (var d2) (bv 8 0x0) false)) ) (float 0 (cast 32 false (>> (var d1) (bv 8 0x0) false)) ))))) (~ (bv 32 0x0)) (bv 32 0x0))) (bv 8 0x0) false)) (set d0 (<< (cast 64 false (ite (&& (! (|| (is_nan (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) )) (is_nan (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) )))) (! (|| (<. (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) ) (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) )) (<. (float 0 (cast 32 false (>> (var d2) (bv 8 0x20) false)) ) (float 0 (cast 32 false (>> (var d1) (bv 8 0x20) false)) ))))) (~ (bv 32 0x0)) (bv 32 0x0))) (bv 8 0x20) false)))
```
And few other instructions.

**Comment by p371jah** [comment]
> Note that it produces different results now:

It is expected since I have simplified expressions.

File: librz/il/il_routines.c [hunk 1]
```diff
@@ -244,3 +244,53 @@ RZ_API RZ_OWN RzILOpBool *rz_il_op_new_ne(RZ_NONNULL RzILOpPure *x, RZ_NONNULL R
...
+	return rz_il_op_new_bool_or(rz_il_op_new_is_nan(x), rz_il_op_new_is_nan(y));
+}
+
+static inline RZ_OWN RzILOpBool *_fl_is_less(RZ_NONNULL RzILOpFloat *x, RZ_NONNULL RzILOpFloat *y) {
+	RzILOpFloat *x_dup = rz_il_op_pure_dup(x);
+	RzILOpFloat *y_dup = rz_il_op_pure_dup(y);
+	return rz_il_op_new_forder(x_dup, y_dup);
+}
+
+RZ_API RZ_OWN RzILOpBool *rz_il_op_new_fneq(RZ_NONNULL RzILOpFloat *x, RZ_NONNULL RzILOpFloat *y) {
```
Comment (Rot127) [inline librz/il/il_routines.c]:
```suggestion
RZ_API RZ_OWN RzILOpBool *rz_il_op_new_fneq(RZ_NONNULL RZ_OWN RzILOpFloat *x, RZ_NONNULL RZ_OWN RzILOpFloat *y) {
```

Also for all the others.

File: librz/il/il_routines.c [hunk 2]
```diff
@@ -244,3 +244,53 @@ RZ_API RZ_OWN RzILOpBool *rz_il_op_new_ne(RZ_NONNULL RzILOpPure *x, RZ_NONNULL R
 	}
 	return rz_il_op_new_bool_inv(result);
 }
+
+static inline RZ_OWN RzILOpBool *_any_fl_is_nan(RZ_NONNULL RZ_OWN RzILOpFloat *x, RZ_NONNULL RZ_OWN RzILOpFloat *y) {
+	return rz_il_op_new_bool_or(rz_il_op_new_is_nan(x), rz_il_op_new_is_nan(y));
+}
+
+static inline RZ_OWN RzILOpBool *_fl_is_less(RZ_NONNULL RzILOpFloat *x, RZ_NONNULL RzILOpFloat *y) {
```
Comment (Rot127) [inline librz/il/il_routines.c]:
```suggestion
static inline RZ_OWN RzILOpBool *_fl_is_less(RZ_NONNULL RZ_BORROW RzILOpFloat *x, RZ_NONNULL RZ_BORROW RzILOpFloat *y) {
```

**Review by Rot127 — CHANGES_REQUESTED** [review]
Had just time for a shallow look. Please check with `valgrind` before merging.
Also, I think it is ok to move it to it's own file `il_fp_routines.c`

**Review by b1llow — APPROVED** [review]

**Comment by p371jah** [comment]
> Please check with `valgrind` before merging.

I use ASAN.

**Review by Heersin — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Pull request merged**